runsc/container/container_test.go

```diff
@@ -3178,3 +3187,156 @@ find $dir -type l -o -type f | sort | xargs cat | md5sum`), 0755); err != nil {
+// the rootfs in container is an EROFS.
+func TestRootfsEROFS(t *testing.T) {
+	// Skip this test if mkfs.erofs or busybox is not available.
+	if _, err := exec.LookPath("mkfs.erofs"); err != nil {
+		t.Skipf("mkfs.erofs is not available: %v", err)
+	}
+	if _, err := exec.LookPath("busybox"); err != nil {
+		t.Skipf("busybox is not available: %v", err)
+	}
+
+	testDir, err := ioutil.TempDir(testutil.TmpDir(), "erofs_rootfs_test_")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir() failed: %v", err)
+	}
+
+	rootfsDir, rootfsImage, err := createRootfsEROFS(testDir)
+	if err != nil {
+		t.Fatalf("failed to create EROFS rootfs image: %v", err)
+	}
+
+	// Create the spec and set the EROFS rootfs annotations.
+	spec := testutil.NewSpecWithArgs("/busybox", "grep", "/ / ro - erofs", "/proc/self/mountinfo")
+	spec.Root.Path = rootfsDir
+	if spec.Annotations == nil {
+		spec.Annotations = make(map[string]string)
+	}
+	spec.Annotations[boot.RootfsPrefix+"type"] = erofs.Name
+	spec.Annotations[boot.RootfsPrefix+"source"] = rootfsImage
+	// Disable the overlay, as we want to be sure that rootfs will always be
+	// shown as EROFS in mountinfo.
+	spec.Annotations[boot.RootfsPrefix+"overlay"] = config.NoOverlay.String()
+
+	conf := testutil.TestConfig(t)
+
+	for _, mounts := range [][]specs.Mount{
+		// Case 1: EROFS rootfs without any other gofer mount.
+		nil,
+
+		// Case 2: EROFS rootfs with a LISAFS backed gofer mount.
+		[]specs.Mount{
+			{
+				Type:        "bind",
+				Destination: "/tmp",
+				Source:      "/tmp",
+			},
+		},
+	} {
+		spec.Mounts = mounts
+
+		_, bundleDir, cleanup, err := testutil.SetupContainer(spec, conf)
+		if err != nil {
+			t.Fatalf("error setting up container: %v", err)
+		}
+		defer cleanup()
+
+		// Create and start the container.
+		args := Args{
+			ID:        testutil.RandomContainerID(),
+			Spec:      spec,
+			BundleDir: bundleDir,
+			Attached:  true,
+		}
+		ws, err := Run(conf, args)
+		if err != nil {
+			t.Fatalf("error running container: %v", err)
+		}
+		if ws.ExitStatus() != 0 {
+			t.Errorf("got exit status %v want %v", ws.ExitStatus(), 0)
+		}
+	}
+}
+
+// TestCheckpointRestoreEROFS does the checkpoint/restore test on each platform using
+// an EROFS image as the rootfs.
+func TestCheckpointRestoreEROFS(t *testing.T) {
+	// Skip this test if mkfs.erofs or busybox is not available.
+	if _, err := exec.LookPath("mkfs.erofs"); err != nil {
+		t.Skipf("mkfs.erofs is not available: %v", err)
+	}
+	if _, err := exec.LookPath("busybox"); err != nil {
+		t.Skipf("busybox is not available: %v", err)
+	}
+
+	testDir, err := ioutil.TempDir(testutil.TmpDir(), "erofs_checkpoint_restore_test_")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir() failed: %v", err)
+	}
+
+	rootfsDir, rootfsImage, err := createRootfsEROFS(testDir)
+	if err != nil {
+		t.Fatalf("failed to create EROFS rootfs image: %v", err)
+	}
+
+	// Skip overlay because test requires writing to host file.
```

Comment:
Comment is stale? We are skipping overlay variants here because they will be ignored. Instead we are setting overlay via annotation..

Reply:
I think the comment is still valid, because it's to describe how overlay is configured for the submount in which the test will create and write the test file:

https://github.com/google/gvisor/blob/42b69d0151b778b53cd23ec7dcca963903857294/pkg/test/testutil/testutil.go#L291-L299

https://github.com/google/gvisor/blob/cae689ed910df140ad3a6e78f7adeed308b57ee6/runsc/container/container_test.go#L1024-L1041